File: dotnet/src/Pactman.NonprofitCheckPlus/Configuration/RetryOptions.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using Pactman.NonprofitCheckPlus.Exceptions;

namespace Pactman.NonprofitCheckPlus.Configuration
{
    /// <summary>
    /// Retry policy. Applied per request, on top of the overall timeout.
    /// </summary>
    /// <remarks>
    /// <para>
    /// Every property carries the SDK default, so <c>new RetryOptions { MaxRetries = 5 }</c>
    /// is a complete policy rather than a patch. To adjust a client's policy for one
    /// request without restating it, use a <c>with</c> expression over the policy in
    /// force: <c>client.Retry with { MaxRetries = 5 }</c>.
    /// </para>
    /// <para>
    /// Values are validated on construction and again on every <c>with</c>, so an
    /// unusable policy fails at the assignment rather than mid-flight on a retry.
    /// </para>
    /// </remarks>
    public sealed record RetryOptions
    {
        /// <summary>Statuses that are never retried, regardless of <see cref="RetryableStatuses"/>.</summary>
        private static readonly int[] NeverRetryStatuses = { 400, 401, 403, 404 };

        private readonly int _maxRetries = 2;
        private readonly TimeSpan _initialDelay = TimeSpan.FromMilliseconds(500);
        private readonly TimeSpan _maxDelay = TimeSpan.FromSeconds(8);
        private readonly double _backoffFactor = 2.0;
        private readonly IReadOnlyList<int> _retryableStatuses = new[] { 429, 500, 502, 503, 504 };

        /// <summary>The SDK's default policy: two retries, jittered exponential backoff.</summary>
        public static RetryOptions Default { get; } = new RetryOptions();

        /// <summary>A policy that never retries.</summary>
        public static RetryOptions None { get; } = new RetryOptions { MaxRetries = 0 };

        /// <summary>
        /// Retries after the first attempt. <c>0</c> disables retrying.
        /// </summary>
        /// <exception cref="PactmanConfigurationException">The value is negative.</exception>
        public int MaxRetries
        {
            get => _maxRetries;
            init
            {
                if (value < 0)
                {
                    throw new PactmanConfigurationException("`MaxRetries` must be 0 or more.");
                }

                _maxRetries = value;
            }
        }

        /// <summary>
        /// Delay before the first retry. Subsequent delays grow by <see cref="BackoffFactor"/>
        /// and are randomized when <see cref="Jitter"/> is on.
        /// </summary>
        /// <exception cref="PactmanConfigurationException">The value is negative.</exception>
        public TimeSpan InitialDelay
        {
            get => _initialDelay;
            init
            {
                if (value < TimeSpan.Zero)
                {
                    throw new PactmanConfigurationException("`InitialDelay` must be 0 or more.");
                }

                _initialDelay = value;
            }
        }

        /// <summary>
        /// Ceiling for a single backoff delay. A server-supplied <c>Retry-After</c> is
        /// honored even when it exceeds this.
        /// </summary>
        /// <exception cref="PactmanConfigurationException">The value is negative.</exception>
        public TimeSpan MaxDelay
        {
            get => _maxDelay;
            init
            {
                if (value < TimeSpan.Zero)
                {
                    throw new PactmanConfigurationException("`MaxDelay` must be 0 or more.");
                }

                _maxDelay = value;
            }
        }

        /// <summary>How much each successive delay grows.</summary>
        /// <exception cref="PactmanConfigurationException">The value is below 1 or not finite.</exception>
        public double BackoffFactor
        {
            get => _backoffFactor;
            init
            {
                if (double.IsNaN(value) || double.IsInfinity(value) || value < 1)
                {
                    throw new PactmanConfigurationException("`BackoffFactor` must be 1 or more.");
                }

                _backoffFactor = value;
            }
        }

        /// <summary>
        /// Randomize each delay across <c>[0, computed]</c> (full jitter) so that clients
        /// failing together do not retry in lockstep.
        /// </summary>
        public bool Jitter { get; init; } = true;

        /// <summary>
        /// HTTP statuses worth retrying.
        /// </summary>
        /// <remarks>
        /// Authentication, authorization, validation and not-found responses are never
        /// retried, whatever this contains.
        /// </remarks>
        /// <exception cref="PactmanConfigurationException">The value is <see langword="null"/>.</exception>
        public IReadOnlyList<int> RetryableStatuses
        {
            get => _retryableStatuses;
            init => _retryableStatuses = value is null
                ? throw new PactmanConfigurationException("`RetryableStatuses` must be a list of status codes.")
                : value.ToArray();
        }

        /// <summary>
        /// Wait for the server's <c>Retry-After</c> before falling back to backoff.
        /// </summary>
        public bool RespectRetryAfter { get; init; } = true;

        /// <summary>True when a status may be retried under this policy.</summary>
        /// <param name="status">The HTTP status the server returned.</param>
        /// <returns><see langword="true"/> when the policy permits another attempt.</returns>
        public bool IsRetryableStatus(int status) =>
            Array.IndexOf(NeverRetryStatuses, status) < 0 && _retryableStatuses.Contains(status);

        /// <summary>A serializable view of the policy.</summary>
        /// <returns>The policy's fields, keyed by name.</returns>
        public IReadOnlyDictionary<string, object?> ToDictionary() =>
            new Dictionary<string, object?>(StringComparer.Ordinal)
            {
                ["maxRetries"] = MaxRetries,
                ["initialDelay"] = InitialDelay.TotalSeconds,
                ["maxDelay"] = MaxDelay.TotalSeconds,
                ["backoffFactor"] = BackoffFactor,
                ["jitter"] = Jitter,
                ["retryableStatuses"] = RetryableStatuses.ToList(),
                ["respectRetryAfter"] = RespectRetryAfter,
            };

        /// <inheritdoc />
        public override string ToString() => string.Format(
            CultureInfo.InvariantCulture,
            "RetryOptions(maxRetries: {0}, initialDelay: {1}s, maxDelay: {2}s, backoffFactor: {3}, jitter: {4})",
            MaxRetries,
            InitialDelay.TotalSeconds,
            MaxDelay.TotalSeconds,
            BackoffFactor,
            Jitter);
    }
}

File: java/sdk/src/main/java/org/pactman/nonprofitcheckplus/config/RetryOptions.java
package org.pactman.nonprofitcheckplus.config;

import java.util.Arrays;
import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.Set;
import org.pactman.nonprofitcheckplus.exceptions.PactmanConfigurationException;

/**
 * Retry policy. Applied per request, on top of the overall timeout.
 *
 * <p>Immutable. Start from {@link #defaults()} or {@link #builder()}, and use
 * {@link #toBuilder()} to change one setting without restating the rest.
 */
public final class RetryOptions {

    /**
     * Statuses that are never retried, whatever {@link #retryableStatuses()}
     * contains.
     *
     * <p>A rejected key, a forbidden resource, a malformed request and a missing
     * record all fail the same way on the second attempt, and retrying them
     * spends quota to learn nothing.
     */
    private static final Set<Integer> NEVER_RETRY_STATUSES =
            Collections.unmodifiableSet(new LinkedHashSet<>(Arrays.asList(400, 401, 403, 404)));

    private static final RetryOptions DEFAULTS = new Builder().build();

    private final int maxRetries;
    private final long initialDelayMs;
    private final long maxDelayMs;
    private final double backoffFactor;
    private final boolean jitter;
    private final Set<Integer> retryableStatuses;
    private final boolean respectRetryAfter;

    private RetryOptions(Builder builder) {
        this.maxRetries = builder.maxRetries;
        this.initialDelayMs = builder.initialDelayMs;
        this.maxDelayMs = builder.maxDelayMs;
        this.backoffFactor = builder.backoffFactor;
        this.jitter = builder.jitter;
        this.retryableStatuses =
                Collections.unmodifiableSet(new LinkedHashSet<>(builder.retryableStatuses));
        this.respectRetryAfter = builder.respectRetryAfter;
    }

    /**
     * The default policy: two retries, 500 ms of backoff growing by two, full
     * jitter, capped at eight seconds.
     *
     * @return the defaults.
     */
    public static RetryOptions defaults() {
        return DEFAULTS;
    }

    /**
     * A policy that never retries.
     *
     * @return the defaults with retrying switched off.
     */
    public static RetryOptions disabled() {
        return DEFAULTS.toBuilder().maxRetries(0).build();
    }

    /**
     * Starts building a policy from the defaults.
     *
     * @return a new builder.
     */
    public static Builder builder() {
        return new Builder();
    }

    /**
     * Starts building a policy from this one.
     *
     * @return a builder holding this policy's settings.
     */
    public Builder toBuilder() {
        return new Builder(this);
    }

    /**
     * Retries after the first attempt. Zero disables retrying.
     *
     * @return the retry count. Defaults to 2, which is three attempts in total.
     */
    public int maxRetries() {
        return maxRetries;
    }

    /**
     * Delay before the first retry.
     *
     * @return the delay in milliseconds. Defaults to 500.
     */
    public long initialDelayMs() {
        return initialDelayMs;
    }

    /**
     * Ceiling for a single backoff delay.
     *
     * <p>A server-supplied {@code Retry-After} is honored even when it exceeds this.
     *
     * @return the ceiling in milliseconds. Defaults to 8000.
     */
    public long maxDelayMs() {
        return maxDelayMs;
    }

    /**
     * How much each delay grows over the last.
     *
     * @return the factor. Defaults to 2.
     */
    public double backoffFactor() {
        return backoffFactor;
    }

    /**
     * Whether each delay is randomized across {@code [0, computed]} (full
     * jitter), so that clients failing together do not retry in lockstep.
     *
     * @return whether jitter is applied. Defaults to true.
     */
    public boolean jitter() {
        return jitter;
    }

    /**
     * HTTP statuses worth retrying.
     *
     * <p>Authentication, authorization, validation and not-found responses are
     * never retried, whatever this contains.
     *
     * @return the statuses. Defaults to 429, 500, 502, 503 and 504.
     */
    public Set<Integer> retryableStatuses() {
        return retryableStatuses;
    }

    /**
     * Whether to wait for the server's {@code Retry-After} before falling back
     * to backoff.
     *
     * @return whether {@code Retry-After} wins. Defaults to true.
     */
    public boolean respectRetryAfter() {
        return respectRetryAfter;
    }

    /**
     * Whether a status may be retried under this policy.
     *
     * @param status the HTTP status code.
     * @return whether another attempt is allowed for this status.
     */
    public boolean isRetryableStatus(int status) {
        return !NEVER_RETRY_STATUSES.contains(status) && retryableStatuses.contains(status);
    }

    /** Collects retry settings. */
    public static final class Builder {
        private int maxRetries = 2;
        private long initialDelayMs = 500L;
        private long maxDelayMs = 8_000L;
        private double backoffFactor = 2.0d;
        private boolean jitter = true;
        private Set<Integer> retryableStatuses =
                new LinkedHashSet<>(Arrays.asList(429, 500, 502, 503, 504));
        private boolean respectRetryAfter = true;

        private Builder() {
        }

        private Builder(RetryOptions source) {
            this.maxRetries = source.maxRetries;
            this.initialDelayMs = source.initialDelayMs;
            this.maxDelayMs = source.maxDelayMs;
            this.backoffFactor = source.backoffFactor;
            this.jitter = source.jitter;
            this.retryableStatuses = new LinkedHashSet<>(source.retryableStatuses);
            this.respectRetryAfter = source.respectRetryAfter;
        }

        /**
         * Sets how many retries follow the first attempt.
         *
         * @param value zero or more.
         * @return this builder.
         */
        public Builder maxRetries(int value) {
            this.maxRetries = value;
            return this;
        }

        /**
         * Sets the delay before the first retry.
         *
         * @param value milliseconds, zero or more.
         * @return this builder.
         */
        public Builder initialDelayMs(long value) {
            this.initialDelayMs = value;
            return this;
        }

        /**
         * Sets the ceiling for a single backoff delay.
         *
         * @param value milliseconds, zero or more.
         * @return this builder.
         */
        public Builder maxDelayMs(long value) {
            this.maxDelayMs = value;
            return this;
        }

        /**
         * Sets how much each delay grows over the last.
         *
         * @param value one or more.
         * @return this builder.
         */
        public Builder backoffFactor(double value) {
            this.backoffFactor = value;
            return this;
        }

        /**
         * Sets whether delays are randomized.
         *
         * @param value whether to apply full jitter.
         * @return this builder.
         */
        public Builder jitter(boolean value) {
            this.jitter = value;
            return this;
        }

        /**
         * Sets which statuses are worth retrying.
         *
         * @param values the statuses. Never-retried statuses stay never-retried.
         * @return this builder.
         */
        public Builder retryableStatuses(Set<Integer> values) {
            this.retryableStatuses = values == null
                    ? new LinkedHashSet<Integer>()
                    : new LinkedHashSet<>(values);
            return this;
        }

        /**
         * Sets whether the server's {@code Retry-After} wins over backoff.
         *
         * @param value whether to honor the header.
         * @return this builder.
         */
        public Builder respectRetryAfter(boolean value) {
            this.respectRetryAfter = value;
            return this;
        }

        /**
         * Validates and builds the policy.
         *
         * @return the immutable policy.
         * @throws PactmanConfigurationException if a setting is nonsensical.
         */
        public RetryOptions build() {
            if (maxRetries < 0) {
                throw new PactmanConfigurationException(
                        "`retry.maxRetries` must be 0 or more.");
            }

            if (initialDelayMs < 0) {
                throw new PactmanConfigurationException(
                        "`retry.initialDelayMs` must be 0 or more.");
            }

            if (maxDelayMs < 0) {
                throw new PactmanConfigurationException("`retry.maxDelayMs` must be 0 or more.");
            }

            if (!(backoffFactor >= 1.0d) || Double.isInfinite(backoffFactor)) {
                throw new PactmanConfigurationException(
                        "`retry.backoffFactor` must be a finite number of 1 or more.");
            }

            return new RetryOptions(this);
        }
    }
}
